roomControl.cs

```diff
@@ -20,122 +21,143 @@ public roomControl()
-           insert.ExecuteNonQuery();
-           connectn.Close();
+                connection.Open();
+                DateTime loginDate;
+                DateTime exitDate;
+
+                string timeQuery = "select  loginDate, exitDate from customer where  ID = @ID ";
+                SqlCommand sqlCommand = new SqlCommand(timeQuery, connection);
+                sqlCommand.Parameters.Add(new System.Data.SqlClient.SqlParameter("@ID", SqlDbType.Int, 3) { Value = convertID });
+                sqlCommand.ExecuteNonQuery();
+
+                SqlDataReader read = sqlCommand.ExecuteReader();
+
+                while (read.Read())
+                {
+                    loginDate = Convert.ToDateTime(read["loginDate"]);
+                    exitDate = Convert.ToDateTime(read["exitDate"]);
+
+                    sumDay = exitDate - loginDate;
+
+                }
+
+                read.Close();
+                connection.Close();
+               
+                connection.Open();
+             
+
+                string insertRoom = "Insert into Room(ID, roomNo, roomColor, howManyDay, checkIn) values ( @ID, @number, Salmon, @howManyDay, GETDATE())";
+                SqlCommand sqlCommandInsert = new SqlCommand(insertRoom, connection);
+                sqlCommandInsert.Parameters.Add(new System.Data.SqlClient.SqlParameter("@ID", SqlDbType.Int, 3) { Value = convertID });
+                sqlCommandInsert.Parameters.Add(new System.Data.SqlClient.SqlParameter("@number", SqlDbType.Int) { Value = number });
+                sqlCommandInsert.Parameters.Add(new System.Data.SqlClient.SqlParameter("@howManyDay", SqlDbType.Int) { Value = sumDay.TotalDays });
+
+                sqlCommandInsert.ExecuteNonQuery();
+                
+                connection.Close();
+            }
         }
 
         private string renk;
         private int RoomNo;
 
-        private void Colorupdate()
+        private void colorUpdate()
         {
-            connectn.Open();
-            Color myColor = Color.Salmon;
+            using (var connection = new SqlConnection(ConnectionString))
+            {
+                connection.Open();
+                Color myColor = Color.Salmon;
 
-            SqlCommand command = new SqlCommand("select roomColor, roomNo from Room  ", connectn);
-            SqlDataReader reader = command.ExecuteReader();
+                SqlCommand command = new SqlCommand("select roomColor, roomNo from Room  ", connection);
+                SqlDataReader reader = command.ExecuteReader();
 
 
-            while (reader.Read())
-            {
+                while (reader.Read())
+                {
 
-                renk = reader["roomColor"].ToString();
-                RoomNo = Convert.ToInt32(reader["roomNo"]);
+                    renk = reader["roomColor"].ToString();
+                    RoomNo = Convert.ToInt32(reader["roomNo"]);
```

Comment:
Int'e cevirmek icin bunu kullanma. Eger bir sorun olursa uygulaman patlar. `TryParse` kullanirsan int disinda gelen degerler icin hata almak yerine false doner. 
```
if (!int.TryParse(reader["roomNo"], out RoomNo))
{
    //Hatali bir oda numarasi geldi eline. Ne yapman gerektigine karar verip kullaniciya hata mesaji donebilirsin.
}
```

Reply:
reader["roomNo"] kısmını bu text ile  kabul etmiyordu. 
genel yazım şekillerini araştırdım. TryParse ilk satırını String olarak aldığı için,
şöyle iki farklı şey denedim. Çalışıyor ancak 2 tane dönüşüm yaptığım için doğruluğundan emin değilim.
               ```
                  string number = reader["roomNo"].ToString();
                    if (!int.TryParse(number, out RoomNo))
                    {
                        MessageBox.Show("error");
                    }

               ```
             Bir başka şekli,
               ```
                       if (!int.TryParse(reader["roomNo"].ToString(), out RoomNo))
                                          {
                                              MessageBox.Show("error");
                                          }
```

Artı olarak ne yapabilirim diye araştırırken de kod kalitesine  güvenmesem de şöyle bir çözüm daha ürettim.
```

                    int i = 1;
                    int? RoomNo;
                    RoomNo = reader.GetInt32(i);
                    if (RoomNo != null)
                    {
                        MessageBox.Show("error");
                    }
                    else 
                    { 
                        i += 2;   // sorgu da 0. --> renk 1.--> oda numarası 2. --> renk 3. --> oda numarası şeklinde devam ettiği için 
                    } 
```

sizin düşüncenizi de merak ediyorum. :)

Reply:
Genelde ikinci yontem yeterli olur. Yalniz dikkat et o alan eger null olabilir bir alan ise null reference exception alabilirsin. Eger roomNo not null bir alan ise direk reader["roomNo"].ToString() yapabilirsin. Null olabilen bir alan ise once null olup olmadigini kontrol etmen lazim. Ucuncu yontem de kullanilabilir ama i tanimlamak yerine yine kolon ismini kullanman daha iyi olur. Bir de GetInt32 kendisi tip donusumu yapacagi icin senin null kontrolu yapman lazim. Yoksa InvalidCastException alirsin. 

Soyle bir sey yapabilirsin:

```
// GetOrdinal -> kolon isminden index getiriyor. Boylece senin index ezberlemene gerek yok, yine kolon ismini kullanabilirsin.
if (!reader.IsDBNull(reader.GetOrdinal("roomNo")))
{
  roomNo = reader.GetInt32(reader.GetOrdinal("roomNo"));
}
else
{
  //room no is null!
}
```

Reply:
anladım, çok teşekkür ederim. Uygulayıp commit edeceğim.